File: Sources/SnapLingo/WorkflowOrchestrator.swift
import Foundation

@MainActor
final class WorkflowOrchestrator {
    private let store: PanelStateStore
    private let capture: SelectionCapture
    private let providerRegistry: ProviderRegistry
    private let panelController: FloatingPanelController
    private var activeTask: Task<Void, Never>?
    private var currentInput: String?

    init(
        store: PanelStateStore,
        capture: SelectionCapture,
        providerRegistry: ProviderRegistry,
        panelController: FloatingPanelController
    ) {
        self.store = store
        self.capture = capture
        self.providerRegistry = providerRegistry
        self.panelController = panelController
    }

    func handleHotkey() async {
        activeTask?.cancel()
        store.resetForNewSession()
        panelController.show()

        activeTask = Task { @MainActor [weak self] in
            guard let self else { return }
            let captureOutcome = await self.capture.captureSelection()
            guard !Task.isCancelled else { return }

            switch captureOutcome {
            case let .text(text):
                await self.process(text: text, sourceLabel: "Auto")
            case .requiresPermission:
                self.store.presentPermissionOnboarding()
            case let .requiresClipboardFallback(changeCount):
                self.store.presentClipboardFallback()
                if let text = await self.capture.waitForClipboardChange(after: changeCount), !Task.isCancelled {
                    await self.process(text: text, sourceLabel: "Clipboard")
                } else if !Task.isCancelled {
                    self.store.showError("SnapLingo never received copied text. Press the hotkey again to retry.")
                }
            }
        }
    }

    func retry() async {
        await handleHotkey()
    }

    func switchMode(to mode: TranslationMode) async {
        guard let currentInput else { return }
        store.selectedMode = mode
        await runProviderPipeline(text: currentInput, selectedMode: mode, modeSourceLabel: "Manual")
    }

    private func process(text: String, sourceLabel: String) async {
        currentInput = text
        let mode = ModeDetector.detect(for: text)
        store.beginProcessing(text: text, selectedMode: mode, modeSourceLabel: sourceLabel)
        await runProviderPipeline(text: text, selectedMode: mode, modeSourceLabel: sourceLabel)
    }

    private func runProviderPipeline(text: String, selectedMode: TranslationMode, modeSourceLabel: String) async {
        store.beginProcessing(text: text, selectedMode: selectedMode, modeSourceLabel: modeSourceLabel)
        let provider = providerRegistry.currentClient()

        do {
            switch selectedMode {
            case .translate:
                let translation = try await provider.translate(text)
                guard !Task.isCancelled else { return }
                store.showPartialTranslation(translation.text)

                let polished = try await provider.polish(translation.text)
                guard !Task.isCancelled else { return }
                store.showFinalResult(title: "Polished Version", text: polished.text)

            case .polish:
                let polished = try await provider.polish(text)
                guard !Task.isCancelled else { return }
                store.showFinalResult(title: "Polished Version", text: polished.text)
            }
        } catch {
            if Task.isCancelled { return }
            let message = (error as? LocalizedError)?.errorDescription ?? error.localizedDescription
            store.showError(message)
        }
    }
}

File: SnapLingoWindows/Services/WorkflowOrchestrator.cs
namespace SnapLingoWindows.Services;

public sealed class WorkflowOrchestrator
{
    private readonly WorkflowStateStore store;
    private readonly ISelectionCaptureService captureService;
    private readonly IProviderClientFactory providerClientFactory;
    private readonly LocalizationService localizer;
    private readonly Action requestPanelPresentation;
    private CancellationTokenSource? activeCts;
    private string? currentInput;

    public WorkflowOrchestrator(
        WorkflowStateStore store,
        ISelectionCaptureService captureService,
        IProviderClientFactory providerClientFactory,
        LocalizationService localizer,
        Action requestPanelPresentation)
    {
        this.store = store;
        this.captureService = captureService;
        this.providerClientFactory = providerClientFactory;
        this.localizer = localizer;
        this.requestPanelPresentation = requestPanelPresentation;
    }

    public async Task HandleHotkeyAsync()
    {
        CancelActiveWork();
        activeCts = new CancellationTokenSource();
        var cancellationToken = activeCts.Token;

        store.ResetForNewSession();
        var capturedText = await captureService.TryCaptureSelectionTextAsync(cancellationToken);
        cancellationToken.ThrowIfCancellationRequested();

        if (!string.IsNullOrWhiteSpace(capturedText))
        {
            requestPanelPresentation();
            await ProcessAsync(capturedText, "source_auto", cancellationToken);
            return;
        }

        var captureOutcome = await captureService.CaptureSelectionAsync(cancellationToken);
        cancellationToken.ThrowIfCancellationRequested();

        if (captureOutcome is not SelectionCaptureOutcome.RequiresClipboardFallback fallback)
        {
            return;
        }

        store.PresentClipboardFallback();
        requestPanelPresentation();
        var clipboardText = await captureService.WaitForClipboardChangeAsync(fallback.ChangeCount, cancellationToken);
        if (!string.IsNullOrWhiteSpace(clipboardText))
        {
            await ProcessAsync(clipboardText, "source_clipboard", cancellationToken);
            return;
        }

        store.ShowError(localizer.Get("error_no_copied_text"));
    }

    public async Task HandleCapturedSelectionAsync(string text, string sourceLabelKey = "source_auto")
    {
        if (string.IsNullOrWhiteSpace(text))
        {
            return;
        }

        CancelActiveWork();
        activeCts = new CancellationTokenSource();
        requestPanelPresentation();
        await ProcessAsync(text.Trim(), sourceLabelKey, activeCts.Token);
    }

    public async Task RetryAsync()
    {
        await HandleHotkeyAsync();
    }

    public async Task SwitchModeAsync(TranslationMode mode)
    {
        if (string.IsNullOrWhiteSpace(currentInput))
        {
            return;
        }

        CancelActiveWork();
        activeCts = new CancellationTokenSource();
        await RunProviderPipelineAsync(currentInput, mode, "Manual", activeCts.Token);
    }

    public async Task UseCurrentClipboardAsync()
    {
        var text = await captureService.ReadClipboardTextAsync();
        if (string.IsNullOrWhiteSpace(text))
        {
            store.ShowError(localizer.Get("error_empty_clipboard"));
            return;
        }

        CancelActiveWork();
        activeCts = new CancellationTokenSource();
        await ProcessAsync(text, "source_clipboard", activeCts.Token);
    }

    private async Task ProcessAsync(string text, string sourceLabelKey, CancellationToken cancellationToken)
    {
        currentInput = text;
        var mode = ModeDetector.Detect(text);
        store.BeginProcessing(text, mode, sourceLabelKey);
        await RunProviderPipelineAsync(text, mode, sourceLabelKey, cancellationToken);
    }

    private async Task RunProviderPipelineAsync(string text, TranslationMode selectedMode, string sourceLabelKey, CancellationToken cancellationToken)
    {
        store.BeginProcessing(text, selectedMode, sourceLabelKey);
        var provider = providerClientFactory.CurrentClient();

        try
        {
            switch (selectedMode)
            {
                case TranslationMode.Translate:
                    var translation = await provider.TranslateAsync(text, cancellationToken);
                    cancellationToken.ThrowIfCancellationRequested();
                    store.ShowPartialTranslation(translation.Text);
                    try
                    {
                        var polishedFromTranslation = await provider.PolishAsync(translation.Text, cancellationToken);
                        cancellationToken.ThrowIfCancellationRequested();
                        store.ShowFinalResult(polishedFromTranslation.Text);
                    }
                    catch (OperationCanceledException)
                    {
                        throw;
                    }
                    catch (Exception)
                    {
                        store.ShowPartialTranslationFallback(localizer.Get("state_partial_kept_status"));
                    }
                    break;

                case TranslationMode.Polish:
                    var polished = await provider.PolishAsync(text, cancellationToken);
                    cancellationToken.ThrowIfCancellationRequested();
                    store.ShowFinalResult(polished.Text);
                    break;
            }
        }
        catch (OperationCanceledException)
        {
        }
        catch (ProviderException error)
        {
            store.ShowError(error.Message);
        }
        catch (Exception error)
        {
            store.ShowError(error.Message);
        }
    }

    private void CancelActiveWork()
    {
        activeCts?.Cancel();
        activeCts?.Dispose();
        activeCts = null;
    }
}
